src/model_vit.py

```diff
@@ -126,6 +129,31 @@ def validation_step(self, batch: torch.Tensor, batch_idx: int) -> torch.Tensor:
         """
         pass
 
+    def predict_step(self, batch: torch.Tensor, batch_idx: int) -> torch.Tensor:
+        """
+        Logic for the neural network's prediction loop.
+        """
+        x: torch.Tensor = batch
+        # x: torch.Tensor = torch.randn(32, 13, 256, 256)  # BCHW
+
+        # Forward encoder
+        outputs_encoder: dict = self(x)
+
+        # Get embeddings generated from encoder
+        embeddings: torch.Tensor = outputs_encoder.last_hidden_state
+        assert embeddings.shape == torch.Size(
+            [self.B, 17, 768]  # (batch_size, sequence_length, hidden_size)
+        )
```

Comment:
@srmsoumya, if you have time, could you try to see why the outputs of the encoder have a shape like `(32, 17, 768)`, or `(batch_size, sequence_length, hidden_size)`? Specifically, I'm not sure what the sequence_length (size: 17) dimension is about, and couldn't quite figure it out from reading https://huggingface.co/docs/transformers/model_doc/vit_mae. More just something to understand, since we'll be moving to your MAE implementation in #47 later.

Reply:
@weiji14 we are masking out `75%` of the patches from the image. Given our image size is `256 x 256` and we use a patch size of `32`, that will give us `256 / 32` => `8 x 8` => `64 patches`. Masking out `75%` of these patches gives us `0.75 x 64 => 16 patches`. Adding 1 extra cls token gives us a total of `17 patches` to input into the transformer portion of the encoder. This results in a batch size of `batch_size: 32`  x `number of unmasked patches: 17` x `patch embedding: 768`. It is also the output we get from the encoder.

When creating the embeddings from the encoder of the model, we should switch-off the masking strategy.

Reply:
Wonderful explanation! Thanks @srmsoumya, I've disabled the masking in the `predict_step` now at commit f09d2e7f45c3d1f30ec7397e4874b26e70fce2a2, and the output is now (32, 65, 768) as expected. Here's a new sample embedding file: [embedding_0.npy.zip](https://github.com/Clay-foundation/model/files/13505741/embedding_0.npy.zip)

Reply:
Hmm, do you have any papers describing when either the cls token or mean of the patch embeddings are used? Also @leothomas, do you have any insights on which method of collapsing the embeddings into a single 1d vector was used for the vector database projects you've worked on? Just want some extra context before decide on which way to go.

Reply:
Ok, after some internal discussion, we decided to go ahead with taking the mean along the spatial dim of the (8x8=64, 768) shaped tensor, to get a final shape of (1, 768). See commit 8eb179da631ffdf87bc4668322d7b13266e424d2, and the Excalidraw illustration in the top post.

Reply:
@danhammer you were asking me this before.

Reply:
I think you are referring to the mode?. I might be wrong, but anyway, it's most probably a small difference in most cases. 

I'm all in to store all 64 at least for one run, or a large sample so we can explore options when we get time. This would clarify then scattering within a tile, which I suspect will not matter for 90% of cases but will matter for strong edges like coast.

Reply:
> Just to clarify, the patches are are 32x32 (pixels) subsections of the overall image, correct? Why are we masking 75% of them? Is that due to cloud cover, or seeking patches only over land?

Yes, each patch are 32x32 pixels. We were masking 75% of the patches because that's how a Masked Autoencoder is trained (see previous PR at #37), but for inference/prediction, we shouldn't apply the mask.

> In regards to the `cls` token, it seems to be a [classification token](https://datascience.stackexchange.com/questions/66207/what-is-purpose-of-the-cls-token-and-why-is-its-encoding-output-important) - which represents the entire image.
> 
> The theory behind averaging the embeddings is that there is a very high liklyhood (but not a guarantee) that the [averages of 2 similar collections will be similar and the averages of 2 difference collections will be different](https://randorithms.com/2020/11/17/Adding-Embeddings.html). It seems that this has mostly been researched in the case of text embeddings, where the exact order of words may matter less than in the case of image patches. I suspect that averaging the embeddings for each patch may lose some of the physical relationships of the overall image, but would be very interesting to compare and contrast the two.
> 
> If we're going to use something like `pgvector`, we can easily have both types of embeddings colocated in the database and build a partial index for each!

That sounds good actually. We could save two files, a cls_token embedding (1x768), and a patch embedding (either 1x768, or 64x768)?

Reply:
If my intuition is correct, is not that surprising that most tiles are "semantically flat" since most tiles will be "on thing" all forest, or grass, or mountain ... The test for the need of richer embeddings would be to pick up examples with semantically rich, or even semantic polarization, like an image with both land and water, or city and forest. 

Also since we train with a MAE, I can see how each semantic patch will actually learn to include the expected semantics of the surrounding patches within the chip, not its own concent; so that the inference can recreate the missing bits, making the stdev within a patch smaller. 

Am I making sense?

Reply:
Yep, it would be good to explore what the embeddings look like for tiles that have more diverse land cover types. I've done just that by picking the most diverse tile we have, details at https://github.com/Clay-foundation/model/issues/35#issuecomment-1837027313

> Also since we train with a MAE, I can see how each semantic patch will actually learn to include the expected semantics of the surrounding patches within the chip, not its own concent; so that the inference can recreate the missing bits, making the stdev within a patch smaller.
> 
> Am I making sense?

Yes, the 1x768 embedding generated from each 32x32 patch actually contains information from the other patches. More details at https://github.com/Clay-foundation/model/discussions/67, and we can continue the discussion there!